File: app/crypto.py
"""Cryptographic functions for VaultUSB."""

import os
import secrets
from pathlib import Path
from typing import Tuple, Optional
import base64
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305
from argon2 import PasswordHasher
from argon2.exceptions import VerifyMismatch
import json
from app.config import config

class CryptoManager:
    """Manages encryption/decryption for the vault."""
    
    def __init__(self):
        self.master_key_file = Path(config.master_key_file)
        self.vault_dir = Path(config.vault_dir)
        self.argon2_hasher = PasswordHasher(
            time_cost=config.argon2_time_cost,
            memory_cost=config.argon2_memory_cost,
            parallelism=config.argon2_parallelism
        )
        self._master_key: Optional[bytes] = None
        self._is_unlocked = False
    
    def generate_master_key(self) -> bytes:
        """Generate a new random master key."""
        return secrets.token_bytes(32)
    
    def derive_key_from_password(self, password: str, salt: bytes) -> bytes:
        """Derive a key from password using Argon2id."""
        return self.argon2_hasher.hash(password.encode()).encode()
    
    def seal_master_key(self, master_key: bytes, password: str) -> str:
        """Seal the master key with a password."""
        salt = secrets.token_bytes(32)
        derived_key = self.derive_key_from_password(password, salt)
        
        # Use first 32 bytes of derived key for encryption
        key = derived_key[:32]
        cipher = ChaCha20Poly1305(key)
        nonce = secrets.token_bytes(12)
        
        sealed_data = cipher.encrypt(nonce, master_key, None)
        
        sealed_info = {
            "salt": base64.b64encode(salt).decode(),
            "nonce": base64.b64encode(nonce).decode(),
            "sealed_data": base64.b64encode(sealed_data).decode()
        }
        
        return json.dumps(sealed_info)
    
    def unseal_master_key(self, sealed_data: str, password: str) -> bytes:
        """Unseal the master key with a password."""
        try:
            sealed_info = json.loads(sealed_data)
            salt = base64.b64decode(sealed_info["salt"])
            nonce = base64.b64decode(sealed_info["nonce"])
            sealed_key = base64.b64decode(sealed_info["sealed_data"])
            
            derived_key = self.derive_key_from_password(password, salt)
            key = derived_key[:32]
            
            cipher = ChaCha20Poly1305(key)
            master_key = cipher.decrypt(nonce, sealed_key, None)
            
            return master_key
        except Exception as e:
            raise ValueError(f"Failed to unseal master key: {e}")
    
    def load_master_key(self, password: str) -> bool:
        """Load and unseal the master key from file."""
        if not self.master_key_file.exists():
            return False
        
        try:
            with open(self.master_key_file, 'r') as f:
                sealed_data = f.read()
            
            self._master_key = self.unseal_master_key(sealed_data, password)
            self._is_unlocked = True
            return True
        except Exception as e:
            print(f"Failed to load master key: {e}")
            return False
    
    def save_master_key(self, master_key: bytes, password: str) -> bool:
        """Save and seal the master key to file."""
        try:
            sealed_data = self.seal_master_key(master_key, password)
            
            # Ensure directory exists
            self.master_key_file.parent.mkdir(parents=True, exist_ok=True)
            
            with open(self.master_key_file, 'w') as f:
                f.write(sealed_data)
            
            return True
        except Exception as e:
            print(f"Failed to save master key: {e}")
            return False
    
    def derive_file_key(self, file_id: str) -> bytes:
        """Derive a file-specific key from master key."""
        if not self._is_unlocked or not self._master_key:
            raise RuntimeError("Master key not unlocked")
        
        hkdf = HKDF(
            algorithm=hashes.SHA256(),
            length=config.file_key_size,
            salt=b"vaultusb_file_key",
            info=file_id.encode()
        )
        return hkdf.derive(self._master_key)
    
    def encrypt_file(self, file_path: Path, file_id: str) -> bool:
        """Encrypt a file in place."""
        if not self._is_unlocked or not self._master_key:
            raise RuntimeError("Master key not unlocked")
        
        try:
            file_key = self.derive_file_key(file_id)
            cipher = ChaCha20Poly1305(file_key)
            
            # Read original file
            with open(file_path, 'rb') as f:
                plaintext = f.read()
            
            # Encrypt
            nonce = secrets.token_bytes(12)
            ciphertext = cipher.encrypt(nonce, plaintext, None)
            
            # Write encrypted file
            with open(file_path, 'wb') as f:
                f.write(nonce + ciphertext)
            
            return True
        except Exception as e:
            print(f"Failed to encrypt file {file_path}: {e}")
            return False
    
    def decrypt_file(self, file_path: Path, file_id: str) -> bytes:
        """Decrypt a file and return plaintext."""
        if not self._is_unlocked or not self._master_key:
            raise RuntimeError("Master key not unlocked")
        
        try:
            file_key = self.derive_file_key(file_id)
            cipher = ChaCha20Poly1305(file_key)
            
            # Read encrypted file
            with open(file_path, 'rb') as f:
                encrypted_data = f.read()
            
            # Split nonce and ciphertext
            nonce = encrypted_data[:12]
            ciphertext = encrypted_data[12:]
            
            # Decrypt
            plaintext = cipher.decrypt(nonce, ciphertext, None)
            
            return plaintext
        except Exception as e:
            raise ValueError(f"Failed to decrypt file {file_path}: {e}")
    
    def secure_delete(self, file_path: Path) -> bool:
        """Securely delete a file by overwriting it with random data."""
        try:
            if not file_path.exists():
                return True
            
            # Get file size
            file_size = file_path.stat().st_size
            
            # Overwrite with random data multiple times
            with open(file_path, 'r+b') as f:
                for _ in range(3):
                    f.seek(0)
                    f.write(secrets.token_bytes(file_size))
                    f.flush()
                    os.fsync(f.fileno())
            
            # Delete the file
            file_path.unlink()
            return True
        except Exception as e:
            print(f"Failed to securely delete {file_path}: {e}")
            return False
    
    def is_unlocked(self) -> bool:
        """Check if the vault is unlocked."""
        return self._is_unlocked and self._master_key is not None
    
    def lock(self):
        """Lock the vault by clearing the master key from memory."""
        self._master_key = None
        self._is_unlocked = False

# Global crypto manager instance
crypto_manager = CryptoManager()

File: cpp/src/crypto.cpp
#include "crypto.h"
#include "config.h"
#include <iostream>
#include <fstream>
#include <random>
#include <sstream>
#include <iomanip>
#include <algorithm>
#include <openssl/evp.h>
#include <openssl/hkdf.h>
#include <openssl/rand.h>
#include <argon2.h>

namespace vaultusb {

CryptoManager& CryptoManager::instance() {
    static CryptoManager instance;
    return instance;
}

CryptoManager::CryptoManager() {
    master_key_file_ = Config::instance().master_key_file();
    vault_dir_ = Config::instance().vault_dir();
    argon2_time_cost_ = Config::instance().argon2_time_cost();
    argon2_memory_cost_ = Config::instance().argon2_memory_cost();
    argon2_parallelism_ = Config::instance().argon2_parallelism();
    file_key_size_ = Config::instance().file_key_size();
}

std::vector<uint8_t> CryptoManager::generate_master_key() {
    std::vector<uint8_t> key(32);
    RAND_bytes(key.data(), key.size());
    return key;
}

std::string CryptoManager::seal_master_key(const std::vector<uint8_t>& master_key, const std::string& password) {
    auto salt = generate_salt();
    auto derived_key = derive_key_from_password(password, salt);
    auto nonce = generate_nonce();
    
    auto encrypted = encrypt_data(master_key, derived_key, nonce);
    
    // Simple JSON-like format
    std::ostringstream json;
    json << "{\"salt\":\"";
    for (auto b : salt) json << std::hex << std::setw(2) << std::setfill('0') << (int)b;
    json << "\",\"nonce\":\"";
    for (auto b : nonce) json << std::hex << std::setw(2) << std::setfill('0') << (int)b;
    json << "\",\"data\":\"";
    for (auto b : encrypted) json << std::hex << std::setw(2) << std::setfill('0') << (int)b;
    json << "\"}";
    
    return json.str();
}

std::vector<uint8_t> CryptoManager::unseal_master_key(const std::string& sealed_data, const std::string& password) {
    // Simple JSON parsing (in production, use a proper JSON library)
    std::string salt_hex, nonce_hex, data_hex;
    
    // Extract hex strings from JSON
    size_t salt_start = sealed_data.find("\"salt\":\"") + 8;
    size_t salt_end = sealed_data.find("\"", salt_start);
    salt_hex = sealed_data.substr(salt_start, salt_end - salt_start);
    
    size_t nonce_start = sealed_data.find("\"nonce\":\"") + 9;
    size_t nonce_end = sealed_data.find("\"", nonce_start);
    nonce_hex = sealed_data.substr(nonce_start, nonce_end - nonce_start);
    
    size_t data_start = sealed_data.find("\"data\":\"") + 8;
    size_t data_end = sealed_data.find("\"", data_start);
    data_hex = sealed_data.substr(data_start, data_end - data_start);
    
    // Convert hex to bytes
    auto salt = hex_to_bytes(salt_hex);
    auto nonce = hex_to_bytes(nonce_hex);
    auto encrypted = hex_to_bytes(data_hex);
    
    auto derived_key = derive_key_from_password(password, salt);
    return decrypt_data(encrypted, derived_key, nonce);
}

bool CryptoManager::load_master_key(const std::string& password) {
    std::ifstream file(master_key_file_);
    if (!file.is_open()) {
        return false;
    }
    
    std::string sealed_data((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
    file.close();
    
    try {
        master_key_ = unseal_master_key(sealed_data, password);
        is_unlocked_ = true;
        return true;
    } catch (const std::exception& e) {
        std::cerr << "Failed to load master key: " << e.what() << std::endl;
        return false;
    }
}

bool CryptoManager::save_master_key(const std::vector<uint8_t>& master_key, const std::string& password) {
    try {
        std::string sealed_data = seal_master_key(master_key, password);
        
        // Ensure directory exists
        std::string dir = master_key_file_.substr(0, master_key_file_.find_last_of('/'));
        system(("mkdir -p " + dir).c_str());
        
        std::ofstream file(master_key_file_);
        if (!file.is_open()) {
            return false;
        }
        
        file << sealed_data;
        file.close();
        
        return true;
    } catch (const std::exception& e) {
        std::cerr << "Failed to save master key: " << e.what() << std::endl;
        return false;
    }
}

std::vector<uint8_t> CryptoManager::derive_file_key(const std::string& file_id) {
    if (!is_unlocked_ || master_key_.empty()) {
        throw std::runtime_error("Master key not unlocked");
    }
    
    return hkdf_derive(master_key_, file_id, file_key_size_);
}

bool CryptoManager::encrypt_file(const std::string& file_path, const std::string& file_id) {
    if (!is_unlocked_ || master_key_.empty()) {
        throw std::runtime_error("Master key not unlocked");
    }
    
    try {
        auto file_key = derive_file_key(file_id);
        auto nonce = generate_nonce();
        
        // Read file
        std::ifstream file(file_path, std::ios::binary);
        if (!file.is_open()) {
            return false;
        }
        
        std::vector<uint8_t> plaintext((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
        file.close();
        
        // Encrypt
        auto encrypted = encrypt_data(plaintext, file_key, nonce);
        
        // Write back (nonce + encrypted data)
        std::ofstream out_file(file_path, std::ios::binary);
        if (!out_file.is_open()) {
            return false;
        }
        
        out_file.write(reinterpret_cast<const char*>(nonce.data()), nonce.size());
        out_file.write(reinterpret_cast<const char*>(encrypted.data()), encrypted.size());
        out_file.close();
        
        return true;
    } catch (const std::exception& e) {
        std::cerr << "Failed to encrypt file: " << e.what() << std::endl;
        return false;
    }
}

std::vector<uint8_t> CryptoManager::decrypt_file(const std::string& file_path, const std::string& file_id) {
    if (!is_unlocked_ || master_key_.empty()) {
        throw std::runtime_error("Master key not unlocked");
    }
    
    try {
        auto file_key = derive_file_key(file_id);
        
        // Read file
        std::ifstream file(file_path, std::ios::binary);
        if (!file.is_open()) {
            throw std::runtime_error("Cannot open file");
        }
        
        std::vector<uint8_t> data((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
        file.close();
        
        if (data.size() < 12) {
            throw std::runtime_error("File too small");
        }
        
        // Split nonce and encrypted data
        std::vector<uint8_t> nonce(data.begin(), data.begin() + 12);
        std::vector<uint8_t> encrypted(data.begin() + 12, data.end());
        
        return decrypt_data(encrypted, file_key, nonce);
    } catch (const std::exception& e) {
        throw std::runtime_error(std::string("Failed to decrypt file: ") + e.what());
    }
}

std::string CryptoManager::hash_password(const std::string& password) {
    std::vector<uint8_t> salt = generate_salt(16);
    std::vector<uint8_t> hash(32);
    
    int result = argon2i_hash_raw(
        argon2_time_cost_, argon2_memory_cost_, argon2_parallelism_,
        password.c_str(), password.length(),
        salt.data(), salt.size(),
        hash.data(), hash.size()
    );
    
    if (result != ARGON2_OK) {
        throw std::runtime_error("Argon2 hash failed");
    }
    
    // Combine salt and hash
    std::vector<uint8_t> combined;
    combined.insert(combined.end(), salt.begin(), salt.end());
    combined.insert(combined.end(), hash.begin(), hash.end());
    
    return bytes_to_hex(combined);
}

bool CryptoManager::verify_password(const std::string& password, const std::string& password_hash) {
    try {
        auto combined = hex_to_bytes(password_hash);
        if (combined.size() < 16) {
            return false;
        }
        
        std::vector<uint8_t> salt(combined.begin(), combined.begin() + 16);
        std::vector<uint8_t> stored_hash(combined.begin() + 16, combined.end());
        std::vector<uint8_t> computed_hash(32);
        
        int result = argon2i_verify(
            stored_hash.data(), stored_hash.size(),
            password.c_str(), password.length(),
            salt.data(), salt.size()
        );
        
        return result == ARGON2_OK;
    } catch (const std::exception&) {
        return false;
    }
}

void CryptoManager::lock() {
    master_key_.clear();
    is_unlocked_ = false;
}

std::vector<uint8_t> CryptoManager::derive_key_from_password(const std::string& password, const std::vector<uint8_t>& salt) {
    std::vector<uint8_t> key(32);
    int result = argon2i_hash_raw(
        argon2_time_cost_, argon2_memory_cost_, argon2_parallelism_,
        password.c_str(), password.length(),
        salt.data(), salt.size(),
        key.data(), key.size()
    );
    
    if (result != ARGON2_OK) {
        throw std::runtime_error("Argon2 key derivation failed");
    }
    
    return key;
}

std::vector<uint8_t> CryptoManager::generate_salt(size_t length) {
    std::vector<uint8_t> salt(length);
    RAND_bytes(salt.data(), salt.size());
    return salt;
}

std::vector<uint8_t> CryptoManager::generate_nonce(size_t length) {
    std::vector<uint8_t> nonce(length);
    RAND_bytes(nonce.data(), nonce.size());
    return nonce;
}

std::vector<uint8_t> CryptoManager::encrypt_data(const std::vector<uint8_t>& data, const std::vector<uint8_t>& key, const std::vector<uint8_t>& nonce) {
    EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
    if (!ctx) {
        throw std::runtime_error("Failed to create cipher context");
    }
    
    if (EVP_EncryptInit_ex(ctx, EVP_chacha20_poly1305(), nullptr, key.data(), nonce.data()) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw std::runtime_error("Failed to initialize encryption");
    }
    
    std::vector<uint8_t> ciphertext(data.size() + 16); // Extra space for tag
    int len;
    int ciphertext_len;
    
    if (EVP_EncryptUpdate(ctx, ciphertext.data(), &len, data.data(), data.size()) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw std::runtime_error("Failed to encrypt data");
    }
    ciphertext_len = len;
    
    if (EVP_EncryptFinal_ex(ctx, ciphertext.data() + len, &len) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw std::runtime_error("Failed to finalize encryption");
    }
    ciphertext_len += len;
    
    EVP_CIPHER_CTX_free(ctx);
    ciphertext.resize(ciphertext_len);
    return ciphertext;
}

std::vector<uint8_t> CryptoManager::decrypt_data(const std::vector<uint8_t>& encrypted_data, const std::vector<uint8_t>& key, const std::vector<uint8_t>& nonce) {
    EVP_CIPHER_CTX* ctx = EVP_CIPHER_CTX_new();
    if (!ctx) {
        throw std::runtime_error("Failed to create cipher context");
    }
    
    if (EVP_DecryptInit_ex(ctx, EVP_chacha20_poly1305(), nullptr, key.data(), nonce.data()) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw std::runtime_error("Failed to initialize decryption");
    }
    
    std::vector<uint8_t> plaintext(encrypted_data.size());
    int len;
    int plaintext_len;
    
    if (EVP_DecryptUpdate(ctx, plaintext.data(), &len, encrypted_data.data(), encrypted_data.size()) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw std::runtime_error("Failed to decrypt data");
    }
    plaintext_len = len;
    
    if (EVP_DecryptFinal_ex(ctx, plaintext.data() + len, &len) != 1) {
        EVP_CIPHER_CTX_free(ctx);
        throw std::runtime_error("Failed to finalize decryption");
    }
    plaintext_len += len;
    
    EVP_CIPHER_CTX_free(ctx);
    plaintext.resize(plaintext_len);
    return plaintext;
}

std::vector<uint8_t> CryptoManager::hkdf_derive(const std::vector<uint8_t>& key, const std::string& info, size_t length) {
    std::vector<uint8_t> derived_key(length);
    
    if (HKDF(derived_key.data(), length, EVP_sha256(), key.data(), key.size(),
             reinterpret_cast<const uint8_t*>("vaultusb_file_key"), 16,
             reinterpret_cast<const uint8_t*>(info.c_str()), info.length()) != 1) {
        throw std::runtime_error("HKDF derivation failed");
    }
    
    return derived_key;
}

std::string CryptoManager::bytes_to_hex(const std::vector<uint8_t>& bytes) {
    std::ostringstream oss;
    for (auto b : bytes) {
        oss << std::hex << std::setw(2) << std::setfill('0') << (int)b;
    }
    return oss.str();
}

std::vector<uint8_t> CryptoManager::hex_to_bytes(const std::string& hex) {
    std::vector<uint8_t> bytes;
    for (size_t i = 0; i < hex.length(); i += 2) {
        std::string byte_str = hex.substr(i, 2);
        bytes.push_back(static_cast<uint8_t>(std::stoul(byte_str, nullptr, 16)));
    }
    return bytes;
}

} // namespace vaultusb
